src/app/3d/qgs3dmaptoolpointcloudchangeattribute.cpp

```diff
@@ -0,0 +1,255 @@
+  {
+    restart();
+  }
+}
+
+void Qgs3DMapToolPointCloudChangeAttribute::mouseReleaseEvent( QMouseEvent *event )
+{
+  if ( ( event->pos() - mClickPoint ).manhattanLength() > QApplication::startDragDistance() )
+    return;
+
+  QgsPoint newPoint = screenPointToMap( event->pos() );
+
+  if ( event->button() == Qt::LeftButton )
+  {
+    if ( mPolygonRubberBand->isEmpty() )
+    {
+      mPolygonRubberBand->addPoint( newPoint );
+      mCanvas->cameraController()->setInputHandlersEnabled( false );
+    }
+    mPolygonRubberBand->addPoint( newPoint );
+    mScreenPoints.append( QgsPoint( event->x(), event->y() ) );
+  }
+  else if ( event->button() == Qt::RightButton )
+  {
+    run();
+    restart();
+  }
+}
+
+void Qgs3DMapToolPointCloudChangeAttribute::activate()
+{
+  // cannot move this to the costructor as there are no mapSettings available yet when the tool is created
+  if ( !mPolygonRubberBand )
+  {
+    mPolygonRubberBand = new QgsRubberBand3D( *mCanvas->mapSettings(), mCanvas->engine(), mCanvas->engine()->frameGraph()->rubberBandsRootEntity(), Qgis::GeometryType::Polygon );
+    mPolygonRubberBand->setHideLastMarker( true );
+  }
+}
+
+void Qgs3DMapToolPointCloudChangeAttribute::deactivate()
+{
+  restart();
+}
+
+void Qgs3DMapToolPointCloudChangeAttribute::setAttribute( const QString &attribute )
+{
+  mAttributeName = attribute;
+}
+
+void Qgs3DMapToolPointCloudChangeAttribute::setNewValue( double value )
+{
+  mNewValue = value;
+}
+
+void Qgs3DMapToolPointCloudChangeAttribute::run()
+{
+  if ( mScreenPoints.size() < 3 )
+    return;
+
+  QgsTemporaryCursorOverride busyCursor( Qt::WaitCursor );
+
+  const QgsGeometry searchPolygon = QgsGeometry( new QgsPolygon( new QgsLineString( mScreenPoints ) ) );
+
+  QgsMapLayer *mapLayer = QgisApp::instance()->activeLayer();
+  Q_ASSERT( mapLayer->type() == Qgis::LayerType::PointCloud );
+  QgsPointCloudLayer *pcLayer = qobject_cast<QgsPointCloudLayer *>( mapLayer );
+  const SelectedPoints sel = searchPoints( pcLayer, searchPolygon );
+
+  int offset;
+  const QgsPointCloudAttribute *attribute = pcLayer->attributes().find( mAttributeName, offset );
+
+  for ( auto it = sel.begin(); it != sel.end(); ++it )
+  {
+    pcLayer->changeAttributeValue( it.key(), it.value(), *attribute, mNewValue );
+  }
+}
+
+void Qgs3DMapToolPointCloudChangeAttribute::restart()
+{
+  mCanvas->cameraController()->setInputHandlersEnabled( true );
+  mScreenPoints.clear();
+  mPolygonRubberBand->reset();
+}
+
+
+QgsPoint Qgs3DMapToolPointCloudChangeAttribute::screenPointToMap( const QPoint &pos ) const
+{
+  const QgsRay3D ray = Qgs3DUtils::rayFromScreenPoint( pos, mCanvas->size(), mCanvas->cameraController()->camera() );
+
+  // pick an arbitrary point mid-way between near and far plane
+  const float pointDistance = ( mCanvas->cameraController()->camera()->farPlane() + mCanvas->cameraController()->camera()->nearPlane() ) / 2;
+  const QVector3D aa = ray.origin() + pointDistance * ray.direction().normalized();
```

Comment:
let's rename `aa` to e.g. `pointWorld` given that it's in the world coordinates? And maybe `newPoint` could be renamed to `pointMap` to make it a bit more clear it's in map coordinates?